lib/AST/ASTWalker.cpp

```diff
@@ -1482,7 +1482,7 @@ class Traversal : public ASTVisitor<Traversal, Expr*, Stmt*,
   
   bool shouldSkip(Decl *D) {
     if (!Walker.shouldWalkMacroArgumentsAndExpansion().second &&
-        D->isInMacroExpansionInContext())
+        D->isInMacroExpansionInContext() && !Walker.Parent.isNull())
```

Comment:
`TypeRefinementContextBuilder` doesn't walk into expansions, but It does `decl->walk(walker)` on the top level decls of the synthesized buffer.